singleobject/reader.go

```diff
@@ -0,0 +1,59 @@
+package singleobject
+
+import (
+	"bytes"
+	"errors"
+)
+
+type soReader struct {
+	r      *bytes.Buffer
+	header []byte
+}
+
+func NewReader(buf *bytes.Buffer) *soReader {
```

Comment:
I know I asked for a reader/writer interface but I realize it probably makes more sense for this method to have a signature like:

```
func ReadHeader(r io.Reader) (fingerprint string, err error)
```

And then users would be able to parse the header and switch on it to figure out what schema it represents like:

```
fingerprint, err := ReadHeader(r)
switch fingerprint {
  case SchemaOneFingerprint:
    DeserializeSchemaOne(r)
  case SchemaTwoFingerprint:
    DeserializeSchemaTwo(r)
}
```

Reply:
did the change code is much simpler